### PR TITLE
Make gradle take build number from env vars and add git build info

### PR DESCRIPTION
**Pull request by Prof-Bloodstone**
1. Gradle will now take `BUILD_NUMBER` env var as version number, and fallback to commit hash. https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html#Server+Build+Properties
2. Server manifest now contains `Git-Commit` and `Git-Branch` entries
  NOTE: There will be different versions when running `./gradlew runDev` (in which case it contains Paper-Server info) vs `./gradlew runShadow` etc.
3. On startup as well as when executing `/version` command, the git information is shown.
  ```
  This server is running Paper version git-Paper-"1706351" (MC: 1.17) (Implementing API version 1.17-R0.1-SNAPSHOT) (Git: build_manifest_version / 1706351)
  ```
  
Other comments will be in my self-review in a moment.

### CODE REVIEW

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 1]
```diff
@@ -0,0 +1,164 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Professor Bloodstone <git@bloodstone.dev>
+Date: Sun, 20 Jun 2021 01:48:31 +0200
+Subject: [PATCH] Add Git information to version command/on startup
```
Comment (Prof-Bloodstone) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: Why new patch?
Initially considered putting it in `Version Command 2.0` patch, but since I not only modify it, I decided to put it in a standalone patch.

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+
+diff --git a/src/main/java/io/papermc/paper/util/JarManifests.java b/src/main/java/io/papermc/paper/util/JarManifests.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..cb3d68b28f66a8559018bd19bc106d136f2fc2f4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/util/JarManifests.java
```
Comment (Prof-Bloodstone) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: If I understand correctly, this should be enough attribution. But maybe I'm wrong?
Should it possibly be a standalone patch with original author?

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
++    /**
++     * Get map containing entries from all manifests
++     *
++     * @throws IOException if unable to get manifest streams
++     */
++    public Map<String, String> getManifestMap() throws IOException {
++        final HashMap<String, String> attribs = new HashMap<>();
++        for (final InputStream stream : getManifestStreams()) {
++            for (final Map.Entry<Object, Object> attr: new Manifest(stream).getMainAttributes().entrySet()) {
++                attribs.put(attr.getKey().toString(), attr.getValue().toString());
```
Comment (Prof-Bloodstone) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: Should we not override the attributes in case of multiple Manifests specifying the same keys?
That's how original code did it: https://github.com/jcabi/jcabi-manifests/blob/c4e1dd22bb6099769b8d279ebe3737e5b638b278/src/main/java/com/jcabi/manifests/Manifests.java#L218-L223

File: patches/server/0714-Add-git-branch-and-commit-to-manifest.patch [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+diff --git a/build.gradle.kts b/build.gradle.kts
+index 0e97fc8f6b961cb28e2338dd101e80fbbc580444..cba95ef8cdbbe8dff48218ef47be0a54e0778dbc 100644
+--- a/build.gradle.kts
++++ b/build.gradle.kts
+@@ -64,6 +64,7 @@ tasks.jar {
+         val git = Git(rootProject.layout.projectDirectory.path)
+         val gitHash = git("rev-parse", "HEAD").getText().substring(0, 7)
+         val implementationVersion = System.getenv("BUILD_NUMBER") ?: "\"$gitHash\""
++        val gitBranch = git("rev-parse", "--abbrev-ref", "HEAD").getText() // Paper
```
Comment (Prof-Bloodstone) [inline patches/server/0714-Add-git-branch-and-commit-to-manifest.patch]: I seem to have lost `.trim()` here. Oops. Will fix in a moment.

File: patches/server/0001-Setup-Gradle-project.patch [hunk 5]
```diff
@@ -71,10 +71,11 @@ index 0000000000000000000000000000000000000000..596b4f52bc57eca8cc7a0138a1e36d71
 +    manifest {
 +        val git = Git(rootProject.layout.projectDirectory.path)
 +        val gitHash = git("rev-parse", "HEAD").getText().substring(0, 7)
++        val implementationVersion = System.getenv("BUILD_NUMBER") ?: "\"$gitHash\""
```
Comment (kashike) [inline patches/server/0001-Setup-Gradle-project.patch:74]: Is this the right environment variable for TeamCity?

Comment (Prof-Bloodstone) [reply patches/server/0001-Setup-Gradle-project.patch:74]: Yes, that's why I linked the TeamCity documentation. Every CI system I've used did use it, but doesn't hurt to make sure :)
Not sure what the build number format is configured on your side, but I imagine its this.
image

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+@@ -85,7 +86,12 @@ public final class Bukkit {
+         }
+ 
+         Bukkit.server = server;
+-        server.getLogger().info("This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ")");
++        // Paper start - add git information
++        Map<String, String> attributes = JarManifests.MANIFEST_MAP;
++        @NotNull String gitBranch = attributes.get("Git-Branch");
++        @NotNull String gitCommit = attributes.get("Git-Commit");
++        server.getLogger().info("This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitBranch + " / " + gitCommit + ")");
```
Comment (kashike) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]:
```suggestion
+        server.getLogger().info("This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitCommit + " on " + gitBranch + ")");
```

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+@@ -240,8 +243,11 @@ public class VersionCommand extends BukkitCommand {
+     // Paper start
+     private void setVersionMessage(final @NotNull Component msg) {
+         lastCheck = System.currentTimeMillis();
++        Map<String, String> attributes = JarManifests.MANIFEST_MAP;
++        @NotNull String gitBranch = attributes.get("Git-Branch");
++        @NotNull String gitCommit = attributes.get("Git-Commit");
+         final Component message = net.kyori.adventure.text.TextComponent.ofChildren(
+-            Component.text("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")", net.kyori.adventure.text.format.NamedTextColor.WHITE),
++            Component.text("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ") (Git: " + gitBranch + " / " + gitCommit + ")", net.kyori.adventure.text.format.NamedTextColor.WHITE),
```
Comment (kashike) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]:
```suggestion
+            Component.text("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ") (Git: " + gitCommit + " on " + gitBranch + ")", net.kyori.adventure.text.format.NamedTextColor.WHITE),
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch (hunk 1 above)
Comment (Proximyst) [reply patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: Just fixup and add yourself as coauthor.

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch (hunk 2 above)
Comment (Proximyst) [reply patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: Yes, this is good.

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch (hunk 1 above)
Comment (Prof-Bloodstone) [reply patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: I meant - since I modify more than just the command, I decided to use a new patch. If you think it belongs in that patch, I don't mind doing that.

**Review by DenWav — CHANGES_REQUESTED** [review]
lgtm

**Review by kashike — APPROVED** [review]

File: patches/api/0318-Add-Git-information-to-version-command-on-startup.patch [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+     /**
+diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+index 4c2ddc722a9dc4011906ad9530b13fa9be1d3ff9..95cd644bf86d650ede40927e722876732f0293cd 100644
+--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
++++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+@@ -10,13 +10,16 @@ import com.google.gson.JsonSyntaxException;
+ import java.io.BufferedReader;
+ import java.io.IOException;
+ import java.net.URL;
++import java.net.URLClassLoader;
```
Comment (Prof-Bloodstone) [inline patches/api/0318-Add-Git-information-to-version-command-on-startup.patch]: Unnecessary import - will remove in a moment.

**Review by DenWav — APPROVED** [review]

**Comment by Prof-Bloodstone** [comment]
With latest changes, running on a test branch results in:
> This server is running Paper version git-Paper-420 (MC: 1.17) (Implementing API version 1.17-R0.1-SNAPSHOT) (Git: 7325700 on build_manifest_version)

Running on `main` or `master` branch results in:
> This server is running Paper version git-Paper-420 (MC: 1.17) (Implementing API version 1.17-R0.1-SNAPSHOT) (Git: b0626db)

Notable change - the message is now accessible publicly under `Bukkit.getVersionMessage()`

File: patches/api/0320-Add-Git-information-to-version-command-on-startup.patch [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
++        // Paper start - add git information
++        server.getLogger().info(getVersionMessage());
++    }
++    /**
++      * Get message describing the version the server is running on.
++      */
++    public static String getVersionMessage() {
++        Map<String, String> attributes = JarManifests.MANIFEST_MAP;
++        @NotNull String gitBranch = attributes.get("Git-Branch");
++        @NotNull String gitCommit = attributes.get("Git-Commit");
```
Comment (Prof-Bloodstone) [inline patches/api/0320-Add-Git-information-to-version-command-on-startup.patch]: These are TECHNICALLY Nullable, but should always exist in paper jars. To be safe I did reverse comparisons below as to avoid potential NPE, but I'm not sure if we want to maybe set those to `unknown` or something, if they are null?
